demo/data-assimilation-4dvar/demo.ipynb

```diff
@@ -0,0 +1,855 @@
+    "        max_iter: Maximum number of optimization iterations\n",
+    "        n_obs: Number of observation times within the assimilation window\n",
+    "\n",
+    "    Returns:\n",
+    "        Optimal analysis state minimizing the 4D-Var cost function\n",
+    "    \"\"\"\n",
+    "\n",
+    "    @jax.jit\n",
+    "    def cost_fn(x0: jnp.ndarray) -> jnp.ndarray:\n",
+    "        \"\"\"Compute the 4D-Var cost function with background and observation terms.\n",
+    "\n",
+    "        Evaluates the standard 4D-Var objective function J = J_b + J_o, where J_b\n",
+    "        penalizes departure from background and J_o penalizes observation misfits\n",
+    "        across the assimilation window.\n",
+    "\n",
+    "        Args:\n",
+    "            x0: Initial state vector for cost evaluation\n",
+    "\n",
+    "        Returns:\n",
+    "            Total cost function value combining background and observation terms\n",
+    "        \"\"\"\n",
+    "        J_b = 0.0\n",
+    "        J_o = 0.0\n",
+    "\n",
+    "        # Background term\n",
+    "        background_diff = x0 - xb\n",
+    "        J_b = 0.5 * jnp.dot(background_diff, B_inv @ background_diff)\n",
+    "\n",
+    "        # Observation terms across assimilation window\n",
+    "        current_state = x0\n",
+    "        obs_diff = y[0] - H @ current_state\n",
+    "        J_o += 0.5 * jnp.dot(obs_diff, R_inv @ obs_diff)\n",
+    "\n",
+    "        for i in range(1, n_obs):\n",
+    "            current_state = M(current_state)[\"result\"][-1]\n",
+    "            obs_diff = y[i] - H @ current_state\n",
+    "            J_o += 0.5 * jnp.dot(obs_diff, R_inv @ obs_diff)\n",
+    "\n",
+    "        return J_b + J_o\n",
+    "\n",
+    "    x0 = scipy.optimize.minimize(\n",
+    "        cost_fn, xb, method=\"L-BFGS-B\", jac=jax.grad(cost_fn)\n",
+    "    ).x\n",
+    "    return x0"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "But wait! Our model is a Tesseract, which is not natively compatible with JAX's automatic differentiation. So we need to wrap the Tesseract in a JAX-compatible function that can be differentiated, by registering a JAX custom derivative for the Tesseract's `apply` method, via [`jax.pure_callback`](https://docs.jax.dev/en/latest/_autosummary/jax.pure_callback.html) and [`jax.custom_vjp`](https://docs.jax.dev/en/latest/_autosummary/jax.custom_vjp.html).\n",
+    "\n",
+    "<div class=admonition note alert alert-warning>\n",
+    "<p class=admonition-title>Note</p>\n",
+    "\n",
+    "While this works in this example, registering Tesseracts by hand is not the recommended way to use Tesseracts in JAX. Instead, you should check out [Tesseract-JAX](https://github.com/pasteurlabs/tesseract-jax), which automatically registers Tesseracts as JAX-compatible functions, allowing you to use them seamlessly in JAX code without needing to manually define callbacks or register custom derivatives.\n",
```

Comment:
Rather than doing this manually + mentioning that tesseract-jax exists, can we port this example to the latter?

Reply:
This is intentional since it's part of the Tesseract Core docs, so should be standalone (and provide line of sight for how you would build this with other frameworks).